core/trino-main/src/main/java/io/trino/sql/planner/EffectivePredicateExtractor.java

```diff
@@ -458,6 +477,32 @@ else if (hasNaN[i]) {
             return domainTranslator.toPredicate(TupleDomain.withColumnDomains(domains.build()).simplify());
         }
 
+        private Boolean containNullValue(Block rowBlock)
+        {
+            Queue<Block> queue = new ArrayDeque<>();
+            queue.add(rowBlock);
+
+            Block currentBlock;
+
+            while (!queue.isEmpty()) {
+                currentBlock = queue.remove();
+                for (int i = 0; i < currentBlock.getPositionCount(); i++) {
+                    if (currentBlock.isNull(i)) {
+                        return true;
+                    }
+                    else {
+                        if (currentBlock.getSingleValueBlock(i) instanceof RowBlock ||
```

Comment:
In general, it's not safe to assume the blocks are of a specific type. You should use the type to get values from the block.

I would write this as follows:


```java
private boolean hasNestedNulls(Type type, Object value)
{
    if (type instanceof RowType) {
        Block container = (Block) value;
        RowType rowType = (RowType) type;
        for (int i = 0; i < rowType.getFields().size(); i++) {
            Type elementType = rowType.getFields().get(i).getType();

            if (container.isNull(i) || elementHasNulls(elementType, container, i)) {
                return true;
            }
        }
    }
    else if (type instanceof ArrayType) {
        Block container = (Block) value;
        ArrayType arrayType = (ArrayType) type;
        Type elementType = arrayType.getElementType();

        for (int i = 0; i < container.getPositionCount(); i++) {
            if (container.isNull(i) || elementHasNulls(elementType, container, i)) {
                return true;
            }
        }
    }

    return false;
}

private boolean elementHasNulls(Type elementType, Block container, int position)
{
    if (elementType instanceof RowType || elementType instanceof ArrayType) {
        Block element = (Block) elementType.getObject(container, position);
        return hasNestedNulls(elementType, element);
    }

    return false;
}
```

on the call side:

```java
if (hasNestedNulls(type, item)) {
    return TRUE_LITERAL;
}
```